GH2078: Expand environment variables in Uri

Kemyke opened the pull request:
This is the second PR after the first was accidentally merged. Here are my thoughts from the previous conversation:

I was thinking about to do this in `ScriptAnalyzer` but it looked quite risky to me. Maybe it is not a good idea to expand the env variables in all lines because we don't know the context. I mean it would replace variables in string, comments, etc. Furthermore this environment variable expansion is only important in the preprocessor directives. In the cake script you can easily access env vars with `EnvironmentVariable` function. But in the preprocess step it is not possible to use these functions or variables of course so my only idea was the environment variable to access various values during the preprocess step. 
That's why i implemented in `UriDirectiveProcessor` so the expansion happens only in URI-s. 

Or is it possible somehow to use variables in this directives? Like `$"#load {repoRoute}/utils.cake"` or `#addin $"nuget:?package=Cake.Some.Nuget&version={someNugetVersion}"`. Would this be a better way?

devlead commented:
> Maybe it is not a good idea to expand the env variables in all lines because we don't know the context.

The below change shouldn't expand on all lines

https://github.com/cake-build/cake/blob/6011c9b67e6d686ef85fa9bc3c7c79e519ed0ab4/src/Cake.Core/Scripting/Analysis/ScriptAnalyzer.cs#L120

```diff
-if (!_lineProcessors.Any(p => p.Process(context, line, out replacement)))
+if (!_lineProcessors.Any(p => p.Process(context, Environment.ExpandEnvironmentVariables(line), out replacement)))
```

Cause it doesn't change the source line, only the line passed to to the directives, if no directive accepts the line it'll just add the source line
https://github.com/cake-build/cake/blob/6011c9b67e6d686ef85fa9bc3c7c79e519ed0ab4/src/Cake.Core/Scripting/Analysis/ScriptAnalyzer.cs#L122
but if the the line is processed any directive then a replacement line is added instead
https://github.com/cake-build/cake/blob/6011c9b67e6d686ef85fa9bc3c7c79e519ed0ab4/src/Cake.Core/Scripting/Analysis/ScriptAnalyzer.cs#L127

We would still need integration tests for existing directives, but it should only be directives that are affected, also expand environment variables will only expand if the environment variable exists.

The reason I'm contemplating this is that it would be cleaner(no change of base classes or method signatures), and make sure it works consistent across directives.

patriksvensson commented on src/Cake.Core/Scripting/Processors/UriDirectiveProcessor.cs:
```diff
@@ -60,14 +60,17 @@ public sealed override bool Process(IScriptAnalyzerContext context, string line,
             Uri uri;
-            if (!Uri.TryCreate(tokens[1].UnQuote(), UriKind.Absolute, out uri))
+
+            var expandedTokens = tokens.Select(t => Environment.ExpandEnvironmentVariables(t));
```
Perhaps we could add an `Expand` method to `ICakeEnvironment` instead of using System.Environment directly? That way it should be pretty straight forward mocking it during tests.

Kemyke replied:
There is the `GetEnvironmentVariables` function in the `ICakeEnvironment`. Wouldn't it be better to use that method and do the variable substitution manually?

patriksvensson replied:
@Kemyke It would be OK to call `Environment.ExpandEnvironmentVariables(t)` from within the `CakeEnvironment` class. Putting it there would give us a way of mocking the behavior in tests.

Kemyke replied:
@patriksvensson  Yeah, i understand, i just look through the `ICakeEnvironment` interface and it looks a little bit redundant to me to have and `ExpandEnvironmentVariable` method and the `GetEnvironmentVariables` dictionary. You could do the mocking with that dictionary as well. 
It was just an idea to keep the `ICakeEnvironment` interface slim. I will do these changes and modify the PR.

Kemyke commented:
> Cause it doesn't change the source line, only the line passed to to the directives, if no directive accepts the line it'll just add the source line

Ok, now i understand and it looks better to me also.

Kemyke commented:
@devlead @patriksvensson i moved the logic to ScriptAnalyzer and changed the unit test. 
If it is looking better for you i will do all the other tests for all directives.

gep13 commented:
@Kemyke was there an issue created to go along with this PR? Thanks

Kemyke commented:
@gep13 there is an improvement tagged issue which is slightly related to this PR i think https://github.com/cake-build/cake/issues/85

gep13 commented:
@Kemyke agreed.  I don't think it would close that issue, but I do agree that it is related. :+1:

Kemyke commented:
@gep13 please let me know if there is anything i can do to move forward with this PR!

devlead commented:
Relates to #2078

devlead commented on src/Cake.Core.Tests/Unit/Scripting/Processors/LoadDirectiveProcessorTests.cs:
```diff
@@ -69,6 +69,25 @@ public void Should_Process_Multiple_Script_References_Found_In_Source(string sou
 
+        [Theory]
+        [InlineData("#l \"%CAKE_TEST_SCRIPT_PATH%/utils.cake\"")]
+        public void Should_Process_Environment_Variable_Found_In_Source(string source)
```
This tests fails on MacOS / Linux

```
[xUnit.net 00:00:02.7407740]     Cake.Core.Tests.Unit.Scripting.Processors.LoadDirectiveProcessorTests.Should_Process_Environment_Variable_Found_In_Source(source: "#l \"%CAKE_TEST_SCRIPT_PATH%/utils.cake\"") [FAIL]
[xUnit.net 00:00:02.7444020]       Assert.Equal() Failure
[xUnit.net 00:00:02.7445470]       Expected: 1
[xUnit.net 00:00:02.7446970]       Actual:   0
[xUnit.net 00:00:02.7463720]       Stack Trace:
[xUnit.net 00:00:02.7488290]         /Users/vagrant/git/src/Cake.Core.Tests/Unit/Scripting/Processors/LoadDirectiveProcessorTests.cs(84,0): at Cake.Core.Tests.Unit.Scripting.Processors.LoadDirectiveProcessorTests.Should_Process_Environment_Variable_Found_In_Source(String source)
Failed   Cake.Core.Tests.Unit.Scripting.Processors.LoadDirectiveProcessorTests.Should_Process_Environment_Variable_Found_In_Source(source: "#l \"%CAKE_TEST_SCRIPT_PATH%/utils.cake\"")
Error Message:
 Assert.Equal() Failure
Expected: 1
Actual:   0
Stack Trace:
   at Cake.Core.Tests.Unit.Scripting.Processors.LoadDirectiveProcessorTests.Should_Process_Environment_Variable_Found_In_Source(String source) in /Users/vagrant/git/src/Cake.Core.Tests/Unit/Scripting/Processors/LoadDirectiveProcessorTests.cs:line 84
```

Kemyke replied:
I fixed the test. Now there is a CodeFactor issue which is not really connected to my PR, i think.

devlead commented on src/Cake.Core/ICakeEnvironment.cs:
```diff
@@ -40,6 +40,13 @@ public interface ICakeEnvironment
+        /// </summary>
+        /// <param name="text">A string containing the names of zero or more environment variables.</param>
+        /// <returns>A string with each environment variable replaced by its value.</returns>
+        string ExpandEnvironmentVariables(string text);
```
Been trying this out, and while I really like the feature, we're really hesitant to introduce a breaking change in Cake.Core 0.27.0 - which adding an interface member is.

Been giving some thought on how not to introduce a breaking change, and only possibility would probably be to switch from an interface member to an extension method, something like

`public static string ExpandEnvironmentVariables(this ICakeEnvironment environment, string text)`

Would make it harder to differentiate between Fake and non fake environments, one naive way would be `if (environment is CakeEnvironment)` or just use the same implementation for both (_that way unit test would run same code as scripts_).

`_environmentVariables` could be replaced with `ICakeEnvironment.GetEnvironmentVariables()` in that method.

Kemyke replied:
Can you give me some insight why is it a breaking change to add a method to an interface? Meanwhile i will check how to do the same with extension method.

devlead replied:
An interface is an contract, any addin / module / assembly inheriting that would break as they would no longer uphold the contract.

Kemyke replied:
I didn't know that the addins can implement the ICakeEnvironment interface. When is this a valid use case?
I thought ICakeEnvironment is something what is used by the addins and a new feature won't be a breaking change. And you have to update the Cake reference in the addin only if you would like to use that new feature.

Kemyke replied:
For the extension method:
what if i don't use the `Environment.ExpandEnvironmentVariables`method but we use the `GetEnvironmentVariables()` of the `ICakeEnvironment` so it will work for normal and fake environment as well. The implementation can be almost the same what is in the `FakeEnvironment.xpandEnvironmentVariables` now.
What do you think? I think this is a better approach than casting the interface to exact types.

devlead replied:
Issue is that it's a public interface, as such we can't know who inherits from it. Someone might replace it with a module or i.e. proxy it within in a addin or their unit tests etc. 

We try really hard to keep both source code and binary compatibility, and that's why we're very careful about changes in Cake.Core, sometimes the feature justifies a breaking change, but in this case I think it's worth the extra effort to not break existing things.

There's some good guidance on the topic provided by the .NET team at:
https://github.com/dotnet/corefx/tree/master/Documentation/coding-guidelines

devlead replied:
> For the extension method:
what if i don't use the Environment.ExpandEnvironmentVariablesmethod but we use the GetEnvironmentVariables() of the ICakeEnvironment so it will work for normal and fake environment as well. The implementation can be almost the same what is in the FakeEnvironment.xpandEnvironmentVariables now.
What do you think? I think this is a better approach than casting the interface to exact types.

Sounds reasonable to me.

Kemyke commented:
@devlead i create the extension method. I create a new `ICakeEnvironmentExtensions.cs` file and put the logic in it. Is it a good place for it?
Please check how you like this implementation!

devlead commented:
@Kemyke your changes have been merged, thanks for your contribution 👍

Kemyke commented:
Thanks!
